### PR TITLE
Score-nn Evaluation

### PR DESCRIPTION
**Pull request by purujitgoyal**
<del> TODO: Add generic configs with evaluation module

### CODE REVIEW

File: structured_prediction_baselines/models/multilabel_classification.py [hunk 1]
```diff
@@ -83,3 +84,99 @@ def get_true_metrics(self, reset: bool = False) -> Dict[str, float]:
...
+    @torch.no_grad()
+    def calculate_metrics(  # type: ignore
+        self,
+        x: Any,
+        labels: torch.Tensor,
+        y_hat: torch.Tensor,
+        buffer: Dict,
+    ) -> None:
+
+        self.map(y_hat, labels)
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/models/multilabel_classification.py:123]: Instead of repeating code, we should call the super's calculate_metrics() in the beginning. This has two reasons behind it,
1. Metrics are volatile, meaning we might comment out or add metrics as need on the parent class. The child class' reliance on the attributes on self which are created in the parent's __init__(), like `self.map` is not safe. We might get errors if we switch off metrics.

2. If we add another metric for MLC, we don't have to add it to two files.

File: structured_prediction_baselines/models/multilabel_classification.py [hunk 2]
```diff
@@ -83,3 +84,99 @@ def get_true_metrics(self, reset: bool = False) -> Dict[str, float]:
...
+
+        # call evaluation_module on distribution and random samples
+        random_samples = self.get_samples(y_hat_n, random=True)
+        distribution_gbi_samples, _ = self.evaluation_module(x, labels, buffer, init_samples=distribution_samples, index=0)
+        # self.distribution_samples_f1(distribution_gbi_samples, labels)
+        random_gbi_samples, _ = self.evaluation_module(x, labels, buffer, init_samples=random_samples, index=1)
+        # self.random_samples_f1(random_gbi_samples, labels)
+
+    def get_true_metrics(self, reset: bool = False) -> Dict[str, float]:
+        metrics = super().get_true_metrics(reset=reset)
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/models/multilabel_classification.py:154]: This is a good practice for metrics.

File: structured_prediction_baselines/modules/loss/inference_net_loss.py [hunk 3]
```diff
@@ -168,9 +168,7 @@ def _get_values(
         )  # if you call this loss, labels cannot be None
 
         if y_cost_aug is None:
-            y_cost_aug = (
-                y_hat if not self.normalize_y else self.normalize(y_hat)
-            )
+            y_cost_aug = torch.zeros_like(y_hat)
```
Comment (dhruvdcoder) [inline structured_prediction_baselines/modules/loss/inference_net_loss.py:171]: This will break things. Any reason for making y_cost_aug zero instead of assigning it the same value as y_hat?

**Review by dhruvdcoder — CHANGES_REQUESTED** [review]
If you are able to run one of the configs with evaluation on, I would suggest adding the generated .json to the tests. Since, evaluation is not a permanent thing, meaning we will only run it on best models, this will lock in the evaluation setup and protect it from breaking.

File: structured_prediction_baselines/modules/loss/inference_net_loss.py (hunk 3 above)
Comment (purujitgoyal) [reply structured_prediction_baselines/modules/loss/inference_net_loss.py:171]: @leejayyoon and I discussed about this, I think it was specific for the case when there was no cost_augmented_layer and we needed inference_net loss from only y_hat.

Can you specify where exactly will it break things?

Comment (dhruvdcoder) [reply structured_prediction_baselines/modules/loss/inference_net_loss.py:171]: Yes. That is the case I am talking about. Now we are not experimenting using that case but why change it to zero.

Comment (dhruvdcoder) [reply structured_prediction_baselines/modules/loss/inference_net_loss.py:171]: I guess it is fine. We can change it.

**Review by dhruvdcoder — APPROVED** [review]

**Review by dhruvdcoder — APPROVED** [review]
